pkg/osvscanner/vendored_libs.go

```diff
@@ -0,0 +1,129 @@
+package osvscanner
+
+import (
+	"bytes"
+	//nolint:gosec
+	"crypto/md5"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/osv-scanner/pkg/osv"
+	"github.com/google/osv-scanner/pkg/reporter"
+)
+
+var (
+	vendoredLibNames = map[string]struct{}{
+		"3rdparty":    {},
+		"dep":         {},
+		"deps":        {},
+		"thirdparty":  {},
+		"third-party": {},
+		"third_party": {},
+		"libs":        {},
+		"external":    {},
+		"externals":   {},
+		"vendor":      {},
+		"vendored":    {},
+	}
+)
+
+func scanDirWithVendoredLibs(r reporter.Reporter, path string) ([]scannedPackage, error) {
+	r.PrintText(fmt.Sprintf("Scanning directory for vendored libs: %s\n", path))
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var packages []scannedPackage
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		libPath := filepath.Join(path, entry.Name())
+
+		r.PrintText(fmt.Sprintf("Scanning potential vendored dir: %s\n", libPath))
+		// TODO: make this a goroutine to parallelize this operation
+		results, err := queryDetermineVersions(libPath)
+		if err != nil {
+			r.PrintText(fmt.Sprintf("Error scanning sub-directory '%s' with error: %v", libPath, err))
+			continue
+		}
+
+		if len(results.Matches) > 0 && results.Matches[0].Score > determineVersionThreshold {
+			match := results.Matches[0]
+			r.PrintText(fmt.Sprintf("Identified %s as %s at %s.\n", libPath, match.RepoInfo.Address, match.RepoInfo.Commit))
+			packages = append(packages, createCommitQueryPackage(match.RepoInfo.Commit, libPath))
+		}
+	}
+
+	return packages, nil
+}
+
+func queryDetermineVersions(repoDir string) (*osv.DetermineVersionResponse, error) {
+	fileExts := []string{
+		".hpp",
+		".h",
+		".hh",
+		".cc",
+		".c",
+		".cpp",
+	}
+
+	var hashes []osv.DetermineVersionHash
+	if err := filepath.Walk(repoDir, func(p string, info fs.FileInfo, err error) error {
+		if info.IsDir() {
+			if _, err := os.Stat(filepath.Join(p, ".git")); err == nil {
+				// Found a git repo, stop here as otherwise we may get duplicated
+				// results with our regular git commit scanning.
+				return filepath.SkipDir
+			}
+			if _, ok := vendoredLibNames[strings.ToLower(info.Name())]; ok {
+				// Ignore nested vendored libraries, as they can cause bad matches.
+				return filepath.SkipDir
+			}
+
+			return nil
+		}
+
+		windowsEnding := []byte("\r\n")
+		unixEnding := []byte("\n")
+
+		for _, ext := range fileExts {
+			if filepath.Ext(p) != ext {
+				continue
+			}
+
+			buf, err := os.ReadFile(p)
+			if err != nil {
+				return err
+			}
+
+			buf = bytes.ReplaceAll(buf, windowsEnding, unixEnding)
```

Comment:
Hmm, are there going to be upstream repos with windows endings as the source of truth? 

I wonder if we need to be doing this normalization on our indexing side also?

Reply:
I haven't thought about this, it could be the case, though I believe the default is LF for git, so this will be pretty rare.

Reply:
Can add some logging into the indexer to see if this is the case.

Reply:
fwiw I feel like that would be safest so you don't have to think about non-line endings being replaced - I'm pretty sure (especially with what we're discussed in the past @another-rex) that shouldn't be the case because it should be the same underlying bytes (and I expect that's what Git sees too?), but at the end of the day I personally would sleep better knowing that there was a straightforward path to normalization...

I don't know too much about this area of things, but for indexing could you actually index both with and without the replacement (and even do that regardless of if a repo is using `\n` or `\r\n`)? that way the downstream wouldn't ever have to actually care right?

Reply:
Sort of? It will double the storage in datastore and increase the indexer runtime, for I don't think much benefit. 

Are you referring to e.g. UTF-8 strings potentially having the bytes `13 10` appearing later on in a e.g. 3 or 4 byte wide character/rune? (I'm not sure if these character exist, but it is a possibility) This will be pretty rare since C or C++ compilers don't commonly support unicode characters in the source code afaik.

Either way I think that can be avoided by adding the same normalization on the indexer side.